### DISCUSSION TITLE
Is it makes any sense to use `ensure_installed` together with `auto_install = true`?

### DISCUSSION THREAD

**Discussion by powerman** [nvim-treesitter/nvim-treesitter#6965] [Q&A] [answered]
At a glance it looks like we should use either one or another, but not both - what's the point in providing list of languages for `ensure_installed` when any new language will be auto-installed anyway?

**Comment by clason** [comment]
For some languages, you need to make sure that the parser is compatible with the queries provided by nvim-treesitter. For those (at least), you need to ensure they are installed at the start. You might also want to bootstrap a list of languages that should always be available (e.g., if you could be offline while editing a buffer for the first time).

**Reply by powerman** [reply]
> For some languages, you need to make sure that the parser is compatible with the queries provided by nvim-treesitter.

You mean in cases when there are more than one parser for a language? But isn't `ensure_installed` contains languages (filetypes), not parsers? I mean, how `ensure_installed` can install different parser if it uses same filetype string to choose a parser to install as `auto_install`? As far as I understood only difference between them is time when installation happens.

**Reply by clason** [reply, marked as answer]
I mean different versions of the same parser, in particular, the ones that are bundled with Neovim. (Auto install is not triggered if a parser already exists, no matter where it comes from.)
